staging/src/k8s.io/client-go/tools/cache/shared_informer.go

```diff
@@ -223,36 +223,63 @@ type SharedIndexInformer interface {
 }
 
-// NewSharedIndexInformer creates a new instance for the listwatcher.
-// The created informer will not do resyncs if the given
-// defaultEventHandlerResyncPeriod is zero.  Otherwise: for each
+// NewSharedIndexInformer creates a new instance for the ListerWatcher and specified Indexers. See
+// NewSharedIndexInformerWithOptions for full details.
+func NewSharedIndexInformer(lw ListerWatcher, exampleObject runtime.Object, defaultEventHandlerResyncPeriod time.Duration, indexers Indexers) SharedIndexInformer {
+	return NewSharedIndexInformerWithOptions(
+		lw,
+		exampleObject,
+		SharedIndexInformerOptions{
+			ResyncPeriod: defaultEventHandlerResyncPeriod,
+			Indexers:     indexers,
+		},
+	)
+}
+
+// NewSharedIndexInformerWithOptions creates a new instance for the ListerWatcher.
+// The created informer will not do resyncs if options.ResyncPeriod is zero.  Otherwise: for each
 // handler that with a non-zero requested resync period, whether added
 // before or after the informer starts, the nominal resync period is
 // the requested resync period rounded up to a multiple of the
 // informer's resync checking period.  Such an informer's resync
 // checking period is established when the informer starts running,
 // and is the maximum of (a) the minimum of the resync periods
 // requested before the informer starts and the
-// defaultEventHandlerResyncPeriod given here and (b) the constant
+// options.ResyncPeriod given here and (b) the constant
 // `minimumResyncPeriod` defined in this file.
-func NewSharedIndexInformer(lw ListerWatcher, exampleObject runtime.Object, defaultEventHandlerResyncPeriod time.Duration, indexers Indexers) SharedIndexInformer {
+func NewSharedIndexInformerWithOptions(lw ListerWatcher, exampleObject runtime.Object, options SharedIndexInformerOptions) SharedIndexInformer {
 	realClock := &clock.RealClock{}
-	sharedIndexInformer := &sharedIndexInformer{
+
+	return &sharedIndexInformer{
+		indexer:                         NewIndexer(DeletionHandlingMetaNamespaceKeyFunc, options.Indexers),
 		processor:                       &sharedProcessor{clock: realClock},
-		indexer:                         NewIndexer(DeletionHandlingMetaNamespaceKeyFunc, indexers),
 		listerWatcher:                   lw,
 		objectType:                      exampleObject,
-		resyncCheckPeriod:               defaultEventHandlerResyncPeriod,
-		defaultEventHandlerResyncPeriod: defaultEventHandlerResyncPeriod,
-		cacheMutationDetector:           NewCacheMutationDetector(fmt.Sprintf("%T", exampleObject)),
+		objectDescription:               options.ObjectDescription,
+		resyncCheckPeriod:               options.ResyncPeriod,
+		defaultEventHandlerResyncPeriod: options.ResyncPeriod,
 		clock:                           realClock,
+		cacheMutationDetector:           NewCacheMutationDetector(fmt.Sprintf("%T", exampleObject)),
 	}
-	return sharedIndexInformer
+}
+
+// SharedIndexInformerOptions configures a sharedIndexInformer.
```

Comment:
What's the cost/benefit to doing this with an exported options struct vs. a functional options approach?

Reply:
Chatted with Jordan. He suggested the struct was better

Reply:
some of bits I was considering:
* options structs are more discoverable, can easily see what all the options are
* they don't expose functions that can twiddle private vars in the object (you could technically call one of these options functions yourself and monkey with private fields in the returned object at weird times in the lifecycle... you deserve pain if you do, but ...)
* variadics introduce edge cases like having the same option multiple times in the list
* options structs are more common in our code base as well

none of those are deal-breaks for variadic options, but they made me lean towards boring options structs

Reply:
Makes sense. Options + variadics make more sense when the thing you want to accomplish is being closed over from the consumer. Re: twiddling private vars - not if you're outside of the package, though, right?

Reply:
> Re: twiddling private vars - not if you're outside of the package, though, right?

you can if an exported options function lets you, e.g.

```go
// make new reflector
reflector := cache.NewNamedReflector(...)
// twiddle private var
cache.ReflectorName("newname")(reflector)
```

Reply:
Oh, I see what you mean. Ew.

Reply:
Although generally speaking I think the Options pattern uses a stand-in struct with no functionality, so I don't know that you could *do* anything with your output.

```go

type Options struct {
    thing string
}

type Option func(*Options)

func WithThing(thing string) Option {
    return func(o *Options) {
         o.thing = thing
    }
}

func NewItemWithOptions(options ...Option) *Item {
    defaults := Options{...}
    for _, opt := range options {
        opt(defaults)
    }
    return &Item{/*use the defaults */}
}
```